### DISCUSSION TITLE
split alphabeta algorithm

### DISCUSSION THREAD

**Discussion by ghost** [official-stockfish/Stockfish#4031] [Q&A] [answered]
i am trying to find the algorithm that split  alphabeta search function for multithreading 

i find two algorithm :

1-Young Brothers Wait Concept
2-Lazy SMP

how stockfish split algorithm for multithreading ?

**Comment by Joachim26** [comment]
... Notably Stockfish 7, released in January 2016, switched from YBW to lazy SMP.

From:
https://www.chessprogramming.org/Lazy_SMP

**Comment by ghost** [comment, marked as answer]
This is my alpha_beta function code , can you explain exactly how to split this function via lazy_SMP ? 

```
int AlphaBeta(int alpha, int beta, int depth)
{

	if(depth==0) {
		return Evaluate(position);
	}

	Move moves[256];
	bestMove = 0;
	int bestScore = MinEval;
	for (int i = 0; i < moveCount; i++)
	{
		position.MakeMove(moves[i]);

		int score;
		int newDepth = depth - 1;

		score = -AlphaBeta(-beta, -alpha,newDepth);

		position.UnmakeMove();


		if (score > bestScore)
		{
			bestMove = moves[i];
			bestScore = score;
			if (score > alpha)
			{
				alpha = score;
				if (score >= beta)
				{
					return score;
				}
			}
		}
	}

	return bestScore;
}
```

please send exactly lazy_SMP algorithm code ,thanks

**Comment by Joachim26** [comment]
I'm too lazy to send you 'exactly lazy_SMP algorithm code' ;-)
But if I were you I would start reading here:
https://github.com/official-stockfish/Stockfish/blob/master/src/thread.cpp
And then continuing somewhere here:
https://github.com/official-stockfish/Stockfish/tree/master/src

It's OPEN SOURCE! Have fun!

**Comment by ghost** [comment]
lazy_SMP algorithm is very difficult , i cant find the algorithm from stockfish source 

 @vondele say the multithreading doesn't reach faster  , is this true ? 
 
 -------------------------------------------------------
 
 _Adding more threads doesn't reach the same answer faster, it finds a better move in (roughly) the same time._

_Originally posted by @vondele in https://github.com/official-stockfish/Stockfish/issues/3987#issuecomment-1102488855_

**Reply by Joachim26** [reply]
1) Yes, since each thread starts searching at the same root node, and
2) Maintainers always telling the truth ;-)
